packages/web-forms/src/components/common/map/useMapBlock.ts

```diff
@@ -0,0 +1,293 @@
+		};
+
+		if (!navigator.geolocation) {
+			currentState.value = STATES.ERROR;
+			errorMessage.value = friendlyError;
+		}
+
+		navigator.geolocation.getCurrentPosition(
+			(position) => {
+				const coords = fromLonLat([position.coords.longitude, position.coords.latitude]);
+				mapInstance
+					?.getView()
+					.animate({ center: coords, zoom: MAX_ZOOM, duration: ANIMATION_TIME });
+				currentState.value = STATES.READY;
+			},
+			() => {
+				currentState.value = STATES.ERROR;
+				errorMessage.value = friendlyError;
+			},
+			{ enableHighAccuracy: true, timeout: GEOLOCATION_TIMEOUT_MS }
+		);
+	};
+
+	const centerFeatureLocation = (feature: Feature<GeometryType>): void => {
+		const geometry = feature.getGeometry();
+		const size = mapInstance?.getSize();
+		const view = mapInstance?.getView();
+
+		if (!geometry || !size?.length || !view) {
+			return;
+		}
+
+		const width = size[0];
+		const height = size[1];
+		let pixelOffsetY = 0;
+		let pixelOffsetX = -50;
+		if (width < SMALL_DEVICE_WIDTH) {
+			pixelOffsetY = 130;
+			pixelOffsetX = 0;
+		}
+
+		view.centerOn(geometry.getExtent(), size, [
+			width / 2 - pixelOffsetX,
+			height / 2 - pixelOffsetY,
+		]);
+	};
+
+	const loadGeometries = (geoJSON: FeatureCollection): void => {
+		if (!mapInstance) {
+			return;
+		}
+
+		currentState.value = STATES.LOADING;
+		selectFeature(undefined);
+		saveFeature(undefined);
+		featuresSource.clear();
+
+		if (!geoJSON.features.length) {
+			mapInstance?.getView().animate({
+				center: DEFAULT_VIEW_CENTER,
+				zoom: MIN_ZOOM,
+				duration: ANIMATION_TIME,
+			});
+			currentState.value = STATES.READY;
+			return;
+		}
+
+		const features = new GeoJSON().readFeatures(geoJSON, {
+			dataProjection: DEFAULT_GEOJSON_PROJECTION,
+			featureProjection: mapInstance.getView().getProjection(),
+		});
+
+		features.forEach((feature) => {
+			if (!feature.get(FEATURE_ID_PROPERTY)) {
+				feature.set(FEATURE_ID_PROPERTY, crypto.randomUUID());
+			}
+		});
+		featuresSource.addFeatures(features);
+		currentState.value = STATES.READY;
+
+		fitToAllFeatures();
+	};
+
+	const selectFeatureByPosition = (position: Pixel): void => {
+		const hitFeatures = mapInstance?.getFeaturesAtPixel(position, {
+			layerFilter: (layer) => layer instanceof WebGLVectorLayer,
+		});
+
+		const featureToSelect = hitFeatures?.length
+			? (hitFeatures[0] as Feature<GeometryType>)
+			: undefined;
+
+		selectFeature(featureToSelect);
+	};
+
+	const selectFeature = (feature?: Feature<GeometryType>) => (selectedFeature.value = feature);
+
+	const saveFeature = (feature?: Feature<GeometryType>) => (savedFeature.value = feature);
+
+	const setSavedByValueProp = (value: string | undefined): void => {
+		if (!value?.length) {
+			return;
+		}
+
+		const featureToSave = featuresSource.getFeatures()?.find((feature) => {
+			return feature.getProperties()?.odk_value === value;
+		}) as Feature<GeometryType>;
```

Comment:
It looks like `getFeatures()` will create a new array. I think `forEachFeature()` is a way to iterate without creating an array. Maybe not a concern, it would only matter if there are many features such that the new array would be large.